src/gt4py/next/program_processors/runners/dace_iterator/utility.py

```diff
@@ -35,6 +35,14 @@ def as_dace_type(type_: ts.ScalarType):
     raise ValueError(f"scalar type {type_} not supported")
 
 
+def as_scalar_type(dtype: str) -> ts.ScalarType:
+    try:
+        kind = getattr(ts.ScalarKind, dtype.upper())
+    except AttributeError:
+        raise ValueError(f"Data type {dtype} not supported.")
+    return ts.ScalarType(kind)
```

Comment:
What do you mean exactly by `checked_cast`? You can always write a custom Typeguard function (see: https://peps.python.org/pep-0647/#typeguard-type) to do a runtime check whose result is also available for the static type checker.

In case you're not aware, more detailed data type definitions were added for the ongoing work with the storages and the embedded mode which could be useful here (but I don't really know because I'm not following closely the PR): https://github.com/GridTools/gt4py/blob/main/src/gt4py/_core/definitions.py  
In the incoming PR with the remaining part of the storages there are some changes and fixes to this file